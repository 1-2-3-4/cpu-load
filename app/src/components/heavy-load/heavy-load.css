.heavy-load-container {
  background-color: #464545;
  border-radius: 4px;
  padding: 25px;
  -webkit-box-shadow: 10px 10px #757575;
          box-shadow: 10px 10px #757575;
  width: 45%;
  margin-left: 5%;
}

.heavy-load-text {
  font-size: 16px;
}

.heavy-load-details-container {
  padding: 10px 20px;
  line-height: 1.4;
  height: 80%;
  background-color: #252525;
  margin: 10% 0;
}

.heavy-load-resolved {
  color: #20b529;
  font-weight: bold;
}

.heavy-load-unresolved {
  color: #d0321d;
  font-weight: bold;
}

.recovery-time-resolved {
  color: #7d7d7d;
}

.recovery-time-unresolved {
  color: #252525;
  background-color: #cc321d;
  padding-left: 10px;
}