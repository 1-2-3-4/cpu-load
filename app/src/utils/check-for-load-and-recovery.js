const HEAVY_CPU_LOAD = 100

export const checkForLoadAndRecovery = (cpuLoad, timeWindow) => {
  const heavyWindowSum = timeWindow * HEAVY_CPU_LOAD
  const heavyLoadTimes = []
  const recoveryTimes = []

  if (cpuLoad.length < timeWindow) {
    return {
      heavyLoadTimes,
      recoveryTimes
    }
  }

  let left = 0
  let tempSum = 0
  let recoveryCounter = 0
  let hasHeavyLoad = false

  /* calculate initial window */
  for (let i = 0; i < timeWindow; i ++) {
    tempSum += cpuLoad[i]
  }

  for (let j = timeWindow; j < cpuLoad.length; j++) {
    /* mark the start time of heavy load */
    if (!hasHeavyLoad && tempSum >= heavyWindowSum) {
      hasHeavyLoad = true
      heavyLoadTimes.push(formatTimeIndex(left))
    }

    /* if there is heavy load, find full window of heavy load */
    while (hasHeavyLoad && recoveryCounter < timeWindow && j < cpuLoad.length) {
      if (cpuLoad[j] < HEAVY_CPU_LOAD) {
        recoveryCounter++
      } else {
        recoveryCounter = 0
      }
      tempSum = tempSum + cpuLoad[j] - cpuLoad[left]
      j++
      left++
    }

    /* if recovery threshhold is met, log recovery time and reset */
    if (recoveryCounter === timeWindow) {
      hasHeavyLoad = false
      recoveryCounter = 0
      recoveryTimes.push(formatTimeIndex(left))
    }

    /* advance window */
    tempSum = tempSum + cpuLoad[j] - cpuLoad[left]
    left++
  }

  return {
    heavyLoadTimes,
    recoveryTimes
  }
}

const formatTimeIndex = (num) => {
  let seconds = num * 10
  let minutes = 0

  while (seconds >= 60) {
    minutes++
    seconds -= 60
  }

  return `${minutes}m ${seconds}s`
}
