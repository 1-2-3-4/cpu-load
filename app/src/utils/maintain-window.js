export const maintainWindow = (arr, length) => {
  if (arr.length > length) {
    arr.shift()
  }
  return arr
}