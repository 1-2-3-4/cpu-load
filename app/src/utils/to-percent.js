export const toPercent = (num) => {
  return `${(num).toFixed(2)}%`
}