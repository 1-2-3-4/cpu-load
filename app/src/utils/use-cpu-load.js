import * as React from 'react'
import { CPULoadContext } from '../components/cpu-load-provider/cpu-load-context'

export const useCPULoad = () => React.useContext(CPULoadContext)