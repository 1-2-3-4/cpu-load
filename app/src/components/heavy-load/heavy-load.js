import * as React from 'react'
import cx from 'classnames'

import { checkForLoadAndRecovery } from '../../utils/check-for-load-and-recovery'
import { useCPULoad } from '../../utils/use-cpu-load'

import './heavy-load.css'

const TEN_SECONDS = 10000
const TWO_MINUTES = 120000
export const TWO_MINUTE_WINDOW = TWO_MINUTES / TEN_SECONDS

export const HeavyLoad = () => {
  const cpuLoad = useCPULoad()
  const { heavyLoadTimes, recoveryTimes } = checkForLoadAndRecovery(cpuLoad, TWO_MINUTE_WINDOW)

  return (
    <div className='heavy-load-container'>
      <div className='heavy-load-text'>Heavy Load and Recovery</div>
      <div className='heavy-load-details-container'>
        {heavyLoadTimes.length === 0 &&
          <div>> No heavy CPU load logged</div>
        }
        {heavyLoadTimes.length > 0 && heavyLoadTimes.map((heavyLoadTime, i) => (
          <React.Fragment key={heavyLoadTime}>
            <div className={cx({
              'heavy-load-resolved': recoveryTimes[i],
              'heavy-load-unresolved': !recoveryTimes[i]
            })}>
              {`> Heavy Load started at ${heavyLoadTime}`}
            </div>
            {recoveryTimes[i] && <div className='recovery-time-resolved'>{`Recovered at ${recoveryTimes[i]}`}</div>}
            {!recoveryTimes[i] && <div className='recovery-time-unresolved'>Issue not resolved</div> }
          </React.Fragment>
        ))}
      </div>
    </div>
  )
}