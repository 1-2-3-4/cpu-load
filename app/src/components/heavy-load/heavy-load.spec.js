import * as React from 'react'
import { shallow } from 'enzyme'

import { HeavyLoad } from './heavy-load'
import * as checkForLoadAndRecoveryUtil from '../../utils/check-for-load-and-recovery'
import * as useCPULoadUtil from '../../utils/use-cpu-load'

describe('HeavyLoad', () => {
  describe('when it renders with no heavy load', () => {
    let wrapper

    beforeAll(() => {
      jest.spyOn(useCPULoadUtil, 'useCPULoad').mockReturnValue([10, 20, 30, 40])
      jest.spyOn(checkForLoadAndRecoveryUtil, 'checkForLoadAndRecovery').mockReturnValue({
        heavyLoadTimes: [],
        recoveryTimes: []
      })

      wrapper = shallow(<HeavyLoad />)
    })

    it('renders `no heavy load` message', () => {
      expect(wrapper.find({ className: 'heavy-load-details-container' }).contains('> No heavy CPU load logged')).toBe(true)
    })

    it('matches snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot()
    })
  })

  describe('when it renders heavy load and no recovery', () => {
    let wrapper

    beforeAll(() => {
      jest.spyOn(useCPULoadUtil, 'useCPULoad').mockReturnValue([100, 100, 100, 100])
      jest.spyOn(checkForLoadAndRecoveryUtil, 'checkForLoadAndRecovery').mockReturnValue({
        heavyLoadTimes: ['1m 30s'],
        recoveryTimes: []
      })

      wrapper = shallow(<HeavyLoad />)
    })

    it('renders a heavy load message', () => {
      expect(wrapper.find({ className: 'heavy-load-unresolved' }).contains('> Heavy Load started at 1m 30s')).toBe(true)
    })

    it('renders a `issue not resolved` message', () => {
      expect(wrapper.find({ className: 'recovery-time-unresolved' }).contains('Issue not resolved')).toBe(true)
    })

    it('matches snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot()
    })
  })

  describe('when it renders with both heavy load and recovery', () => {
    let wrapper

    beforeAll(() => {
      jest.spyOn(useCPULoadUtil, 'useCPULoad').mockReturnValue([100, 100, 100, 100, 20, 20, 20, 20])
      jest.spyOn(checkForLoadAndRecoveryUtil, 'checkForLoadAndRecovery').mockReturnValue({
        heavyLoadTimes: ['1m 30s'],
        recoveryTimes: ['5m 30s']
      })

      wrapper = shallow(<HeavyLoad />)
    })

    it('renders a heavy load message', () => {
      expect(wrapper.find({ className: 'heavy-load-resolved' }).contains('> Heavy Load started at 1m 30s')).toBe(true)
    })

    it('renders a recovery message', () => {
      expect(wrapper.find({ className: 'recovery-time-resolved' }).contains('Recovered at 5m 30s')).toBe(true)
    })

    it('matches snapshot', () => {
      expect(wrapper.getElement()).toMatchSnapshot()
    })
  })
})