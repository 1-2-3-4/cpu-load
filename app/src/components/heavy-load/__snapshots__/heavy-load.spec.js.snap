// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeavyLoad when it renders heavy load and no recovery matches snapshot 1`] = `
<div
  className="heavy-load-container"
>
  <div
    className="heavy-load-text"
  >
    Heavy Load and Recovery
  </div>
  <div
    className="heavy-load-details-container"
  >
    <React.Fragment>
      <div
        className="heavy-load-unresolved"
      >
        &gt; Heavy Load started at 1m 30s
      </div>
      <div
        className="recovery-time-unresolved"
      >
        Issue not resolved
      </div>
    </React.Fragment>
  </div>
</div>
`;

exports[`HeavyLoad when it renders with both heavy load and recovery matches snapshot 1`] = `
<div
  className="heavy-load-container"
>
  <div
    className="heavy-load-text"
  >
    Heavy Load and Recovery
  </div>
  <div
    className="heavy-load-details-container"
  >
    <React.Fragment>
      <div
        className="heavy-load-resolved"
      >
        &gt; Heavy Load started at 1m 30s
      </div>
      <div
        className="recovery-time-resolved"
      >
        Recovered at 5m 30s
      </div>
    </React.Fragment>
  </div>
</div>
`;

exports[`HeavyLoad when it renders with no heavy load matches snapshot 1`] = `
<div
  className="heavy-load-container"
>
  <div
    className="heavy-load-text"
  >
    Heavy Load and Recovery
  </div>
  <div
    className="heavy-load-details-container"
  >
    <div>
      &gt; No heavy CPU load logged
    </div>
  </div>
</div>
`;
